### PR TITLE
[Raiden Weight Sync 6/6] MaxText trainer config plumbing and MoE dimension padding

### PR DESCRIPTION
**Pull request by YixuanWang-99**
## Overview
Part of the stacked Raiden weight-sync enablement PR chain replacing #2083.

**Stack:**
- [T1] https://github.com/google/tunix/pull/2146: Multi-host discovery & multi-axis sharding in RaidenHandler
- [T2a] https://github.com/google/tunix/pull/2147: Transport selection, FFI preflight mismatch check, and host-stage normalization
- [T3] https://github.com/google/tunix/pull/2148: Extract RaidenDestinationWeightSyncMixin
- [T4] https://github.com/google/tunix/pull/2149: Cache lifecycle and prefix caching
- [T5] https://github.com/google/tunix/pull/2150: MoE 128-lane interleaving for TPU GMM layout (pairs with MaxText M2)
- **[T6] google/tunix (this PR)**: MaxText trainer configuration plumbing

## Details
- Plumbs `base_num_kv_heads`, `rollout_mesh_tp`, and `prefuse_moe_weights` into `build_maxtext_config`.
- Automatically computes `padded_moe_mlp_dim` to satisfy TPU kernel constraints.
- Replaces ad-hoc `os.environ` reads with explicit arguments.

### CODE REVIEW

File: tunix/utils/maxtext_utils.py
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py:1]: It seems there is no tests for any of these utils. can you add tests for the changes? (can be a follow up cl)

File: tunix/utils/maxtext_utils.py [hunk 1]
```diff
@@ -59,7 +59,11 @@ def build_maxtext_config(
     warmup_steps_fraction: float = 0.0,
     load_parameters_path: str = "",
     padded_moe_mlp_dim: int = 0,
+    base_num_kv_heads: int = 0,
```
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: do you want to add validation on acceptable ranges?

File: tunix/utils/maxtext_utils.py [hunk 2]
```diff
@@ -86,13 +90,25 @@ def build_maxtext_config(
...
+  if not padded_moe_mlp_dim and rollout_mesh_tp > 0:
+    try:
+      from maxtext.integration.vllm.moe_padding import compute_padded_moe_mlp_dim
+      tmp_cfg = pyconfig.initialize(argv)
+      base_dim = getattr(tmp_cfg, "base_moe_mlp_dim", None) or getattr(tmp_cfg, "moe_intermediate_size", None)
+      if base_dim:
+        padded_moe_mlp_dim = compute_padded_moe_mlp_dim(base_dim, rollout_mesh_tp)
+        logging.info("Auto-computed padded_base_moe_mlp_dim=%d for rollout_mesh_tp=%d", padded_moe_mlp_dim, rollout_mesh_tp)
+    except Exception as e:
+      logging.warning("Could not auto-compute padded_base_moe_mlp_dim: %s", e)
```
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: is it okay to just log a warning? wouldn't this cause a silent failure?

File: tunix/utils/maxtext_utils.py (hunk 1 above)
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: you are adding positional arguments and changing the order. wouldn't this break any existing code?

**Review by s-noghabi — CHANGES_REQUESTED** [review]

**Review by tianshub — APPROVED** [review]
Please see Shadi's comment

File: tunix/utils/maxtext_utils.py (hunk 1 above)
Comment (YixuanWang-99) [reply tunix/utils/maxtext_utils.py]: Great point! We preserved `base_output_directory: str = ""` in its exact original positional order and marked all newly added parameters (`base_num_kv_heads`, `rollout_mesh_tp`, `prefuse_moe_weights`, `use_weight_converter`) as keyword-only using `*,`.

Comment (YixuanWang-99) [reply tunix/utils/maxtext_utils.py]: Added! We now validate that `padded_moe_mlp_dim >= 0`, `base_num_kv_heads >= 0`, and `rollout_mesh_tp >= 0`, raising a clear `ValueError` if negative.

File: tunix/utils/maxtext_utils.py (hunk 2 above)
Comment (YixuanWang-99) [reply tunix/utils/maxtext_utils.py]: Agreed — replaced the warning with `raise RuntimeError(...) from e` to fail fast and prevent silent misconfigurations.

File: tunix/utils/maxtext_utils.py
Comment (YixuanWang-99) [reply tunix/utils/maxtext_utils.py:1]: Added unit tests in `tests/utils/maxtext_utils_test.py` covering:
- Custom hyperparameter propagation and single `pyconfig.initialize` invocation
- Automatic padded MoE MLP dimension computation
- Non-negative range validations
- Auto-padding failure fast-fail behavior
- Batch size divisibility validation.

Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: not in scope of this PR, but shouldn't this be its own independent knob?

**Review by s-noghabi — APPROVED** [review]

File: tunix/utils/maxtext_utils.py [hunk 3]
```diff
@@ -62,14 +66,47 @@ def build_maxtext_config(
...
+    use_weight_converter: bool = True,
 ) -> Any:
   """Builds the MaxText HyperParameters the training engine runs on."""
   pyconfig, _, _ = maxtext_modules()
 
+  # Backward compatibility: if rollout_mesh_tp was provided, default kv_tp_size and moe_mlp_tp_size
+  if rollout_mesh_tp > 0:
+    if kv_tp_size == 0:
+      kv_tp_size = rollout_mesh_tp
+    if moe_mlp_tp_size == 0:
```
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py:83]: add logging for these overrides so it is not a surprise.

File: tunix/utils/maxtext_utils.py [hunk 4]
```diff
@@ -79,6 +116,71 @@ def build_maxtext_config(
...
+    if os.path.exists(model_yml):
+      try:
+        import yaml
+
+        with open(model_yml, "r") as f:
+          data = yaml.safe_load(f)
+        if isinstance(data, dict):
+          effective_kv_heads = int(data.get("base_num_kv_heads") or 0)
+      except Exception:
+        effective_kv_heads = 0
```
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: log that it didn't load the file

File: tunix/utils/maxtext_utils.py [hunk 5]
```diff
@@ -79,6 +116,71 @@ def build_maxtext_config(
...
+        try:
+          import yaml
+
+          with open(model_yml, "r") as f:
+            data = yaml.safe_load(f)
+          if isinstance(data, dict):
+            base_dim = data.get("base_moe_mlp_dim") or data.get(
+                "moe_intermediate_size"
+            )
+        except Exception:
```
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: this seems very similar to above. can you extract all needed config in one file load?

File: tunix/utils/maxtext_utils.py [hunk 6]
```diff
@@ -79,6 +116,71 @@ def build_maxtext_config(
...
+          base_dim = None
```
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: ditto on the error message.

File: tunix/utils/maxtext_utils.py (hunk 4 above)
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: if we set effective_kv_heads = 0 just because we didn't load the file, isn't this a wrong path? shouldn't you raise error if kv_tp_size > 0?

**Review by s-noghabi — CHANGES_REQUESTED** [review]

File: tunix/utils/maxtext_utils.py (hunk 4 above)
Comment (YixuanWang-99) [reply tunix/utils/maxtext_utils.py]: Added

File: tunix/utils/maxtext_utils.py (hunk 3 above)
Comment (YixuanWang-99) [reply tunix/utils/maxtext_utils.py:83]: added

File: tunix/utils/maxtext_utils.py [hunk 7]
```diff
@@ -79,6 +119,83 @@ def build_maxtext_config(
...
+      )
+    effective_kv_heads = kv_tp_size
+
+  # 2. Resolve padded MoE MLP dimension before pyconfig initialization:
+  if not effective_padded_moe_mlp_dim and moe_mlp_tp_size > 0:
+    compute_padded_moe_mlp_dim = None
+    try:
+      from maxtext.integration.vllm.convert_utils import compute_padded_moe_mlp_dim
+    except (ImportError, ModuleNotFoundError):
+      pass
```
Comment (s-noghabi) [inline tunix/utils/maxtext_utils.py]: add warning this is skipped

**Review by s-noghabi — APPROVED** [review]

File: tunix/utils/maxtext_utils.py (hunk 7 above)
Comment (YixuanWang-99) [reply tunix/utils/maxtext_utils.py]: Added warning when compute_padded_moe_mlp_dim cannot be imported, and added test coverage in tests/utils/maxtext_utils_test.py.
